File: windows/tests/ConnectorControl.Core.Tests/PlaceholderTests.cs
namespace ConnectorControl.Core.Tests;

/// <summary>Mirror: Tests/ConnectorControlCoreTests/PlaceholderTests.swift</summary>
public class PlaceholderTests
{
    [Fact]
    public void MarkerSyntax()
    {
        Assert.Equal("${CC_NEEDS:DBT_TOKEN}", Placeholder.Marker("DBT_TOKEN"));
        Assert.True(Placeholder.IsValidName("server_path2"));
        Assert.False(Placeholder.IsValidName("bad-name"));
        Assert.False(Placeholder.IsValidName(""));
    }

    [Fact]
    public void FindsNamesInOrderWithoutDuplicates()
    {
        Assert.Equal(["token", "x"], Placeholder.NamesIn("Bearer ${CC_NEEDS:token} ${CC_NEEDS:token} ${CC_NEEDS:x}"));
        Assert.Empty(Placeholder.NamesIn("${CC_NEEDS:bad-name} plain"));
        Assert.True(Placeholder.ContainsMarker("a ${CC_NEEDS:b} c"));
        Assert.False(Placeholder.ContainsMarker("${COLLECTION_DIR}/x"));
    }

    [Fact]
    public void MarkersInAConfigReportPointers()
    {
        var config = JsonValue.Object(
            ("args", JsonValue.Array([JsonValue.String("${CC_NEEDS:server_path}")])),
            ("env", JsonValue.Object(("AUTH_HEADER", JsonValue.String("Bearer ${CC_NEEDS:token}")), ("PLAIN", JsonValue.String("v")))));
        var found = Placeholder.MarkersIn(config);
        Assert.Equal(["/args/0", "/env/AUTH_HEADER"], found.Select(m => m.Pointer.ToString()).ToArray());
        Assert.Equal([["server_path"], ["token"]], found.Select(m => m.Names.ToArray()).ToArray());
    }

    /// <summary>
    /// Flattened in the order the leaves are walked, so a sentence built from these names reads
    /// the same wherever it is built and however often a name appears.
    /// </summary>
    [Fact]
    public void UnfilledNamesFlattenTheMarkersInFirstAppearanceOrder()
    {
        var config = JsonValue.Object(
            ("args", JsonValue.Array([JsonValue.String("${CC_NEEDS:zulu}"), JsonValue.String("${CC_NEEDS:zulu} ${CC_NEEDS:alpha}")])),
            ("env", JsonValue.Object(("A", JsonValue.String("${CC_NEEDS:alpha}")), ("B", JsonValue.String("plain")))));
        Assert.Equal(["zulu", "alpha"], Placeholder.UnfilledNamesIn(config));
        Assert.Empty(Placeholder.UnfilledNamesIn(JsonValue.Object(("a", JsonValue.String("none")))));
    }

    [Fact]
    public void ExpandsTheDirectoryTokenInEveryStringLeaf()
    {
        var config = JsonValue.Object(
            ("command", JsonValue.String("node")),
            ("args", JsonValue.Array([JsonValue.String("${COLLECTION_DIR}/../servers/x.js")])),
            ("env", JsonValue.Object(("ROOT", JsonValue.String("${COLLECTION_DIR}")))));
        Assert.True(Placeholder.UsesDirectoryToken(config));
        var expanded = Placeholder.ExpandDirectoryToken(config, @"C:\Users\d\Acme\mcp");
        Assert.Equal(JsonValue.String(@"C:\Users\d\Acme\mcp/../servers/x.js"), expanded.ValueAt(new JsonPointer(["args", "0"])));
        Assert.Equal(JsonValue.String(@"C:\Users\d\Acme\mcp"), expanded.ValueAt(new JsonPointer(["env", "ROOT"])));
        Assert.False(Placeholder.UsesDirectoryToken(expanded));
    }

    [Fact]
    public void CollapsesTheDirectoryOnlyWhereItStandsAsAFolder()
    {
        const string share = @"C:\Users\d\share";
        var config = JsonValue.Object(
            ("command", JsonValue.String("node")),
            ("args", JsonValue.Array([
                JsonValue.String(share + "/tools/x.js"), JsonValue.String("--root=" + share),
                JsonValue.String(share + "-tools/y.js"), JsonValue.String(@"\\nas\" + share + @"\z.js"),
            ])),
            ("env", JsonValue.Object(("ROOT", JsonValue.String(share)))));
        var collapsed = Placeholder.CollapseDirectory(config, share);
        // A sibling that begins with the name, and a longer path that holds it, are left alone.
        Assert.Equal(
            ["${COLLECTION_DIR}/tools/x.js", "--root=${COLLECTION_DIR}", share + "-tools/y.js", @"\\nas\" + share + @"\z.js"],
            collapsed.ValueAt(new JsonPointer(["args"]))!.ArrayItems.Select(a => a.StringValue));
        Assert.Equal(JsonValue.String("${COLLECTION_DIR}"), collapsed.ValueAt(new JsonPointer(["env", "ROOT"])));
        // Collapsing undoes what expanding did.
        var one = JsonValue.Object(("args", JsonValue.Array([JsonValue.String(share + "/tools/x.js")])));
        Assert.Equal(one, Placeholder.ExpandDirectoryToken(Placeholder.CollapseDirectory(one, share), share));
        Assert.Equal(config, Placeholder.CollapseDirectory(config, ""));
    }
}

File: Tests/ConnectorControlCoreTests/PlaceholderTests.swift
import XCTest
import ConnectorControlTestSupport
@testable import ConnectorControlCore

/// Mirror: windows/tests/ConnectorControl.Core.Tests/PlaceholderTests.cs
final class PlaceholderTests: XCTestCase {
    func testMarkerSyntax() {
        XCTAssertEqual(Placeholder.marker("DBT_TOKEN"), "${CC_NEEDS:DBT_TOKEN}")
        XCTAssertTrue(Placeholder.isValidName("server_path2"))
        XCTAssertFalse(Placeholder.isValidName("bad-name"))
        XCTAssertFalse(Placeholder.isValidName(""))
    }
    func testFindsNamesInOrderWithoutDuplicates() {
        XCTAssertEqual(Placeholder.names(in: "Bearer ${CC_NEEDS:token} ${CC_NEEDS:token} ${CC_NEEDS:x}"), ["token", "x"])
        XCTAssertEqual(Placeholder.names(in: "${CC_NEEDS:bad-name} plain"), [])
        XCTAssertTrue(Placeholder.containsMarker("a ${CC_NEEDS:b} c"))
        XCTAssertFalse(Placeholder.containsMarker("${COLLECTION_DIR}/x"))
    }
    func testMarkersInAConfigReportPointers() {
        let config: JSONValue = .object([
            "args": .array([.string("${CC_NEEDS:server_path}")]),
            "env": .object(["AUTH_HEADER": .string("Bearer ${CC_NEEDS:token}"), "PLAIN": .string("v")]),
        ])
        let found = Placeholder.markers(in: config)
        XCTAssertEqual(found.map { $0.pointer.description }, ["/args/0", "/env/AUTH_HEADER"])
        XCTAssertEqual(found.map { $0.names }, [["server_path"], ["token"]])
    }

    /// Flattened in the order the leaves are walked, so a sentence built from these names reads
    /// the same wherever it is built and however often a name appears.
    func testUnfilledNamesFlattenTheMarkersInFirstAppearanceOrder() {
        let config: JSONValue = .object([
            "args": .array([.string("${CC_NEEDS:zulu}"), .string("${CC_NEEDS:zulu} ${CC_NEEDS:alpha}")]),
            "env": .object(["A": .string("${CC_NEEDS:alpha}"), "B": .string("plain")]),
        ])
        XCTAssertEqual(Placeholder.unfilledNames(in: config), ["zulu", "alpha"])
        XCTAssertEqual(Placeholder.unfilledNames(in: .object(["a": .string("none")])), [])
    }
    func testExpandsTheDirectoryTokenInEveryStringLeaf() {
        let config: JSONValue = .object([
            "command": .string("node"),
            "args": .array([.string("${COLLECTION_DIR}/../servers/x.js")]),
            "env": .object(["ROOT": .string("${COLLECTION_DIR}")]),
        ])
        XCTAssertTrue(Placeholder.usesDirectoryToken(config))
        let expanded = Placeholder.expandDirectoryToken(in: config, directory: "/Users/d/Acme/mcp")
        XCTAssertEqual(expanded.value(at: JSONPointer(["args", "0"])), .string("/Users/d/Acme/mcp/../servers/x.js"))
        XCTAssertEqual(expanded.value(at: JSONPointer(["env", "ROOT"])), .string("/Users/d/Acme/mcp"))
        XCTAssertFalse(Placeholder.usesDirectoryToken(expanded))
    }

    func testCollapsesTheDirectoryOnlyWhereItStandsAsAFolder() {
        let config: JSONValue = .object([
            "command": .string("node"),
            "args": .array([.string("/Users/d/share/tools/x.js"), .string("--root=/Users/d/share"),
                            .string("/Users/d/share-tools/y.js"), .string("/home/Users/d/share/z.js")]),
            "env": .object(["ROOT": .string("/Users/d/share")]),
        ])
        let collapsed = Placeholder.collapseDirectory(in: config, directory: "/Users/d/share")
        XCTAssertEqual(args(of: collapsed), ["${COLLECTION_DIR}/tools/x.js", "--root=${COLLECTION_DIR}",
                                             "/Users/d/share-tools/y.js", "/home/Users/d/share/z.js"],
                       "a sibling that begins with the name, and a longer path that ends with it, are left alone")
        XCTAssertEqual(collapsed.value(at: JSONPointer(["env", "ROOT"])), .string("${COLLECTION_DIR}"))
        XCTAssertEqual(Placeholder.expandDirectoryToken(in: Placeholder.collapseDirectory(
            in: .object(["args": .array([.string("/Users/d/share/tools/x.js")])]), directory: "/Users/d/share"),
            directory: "/Users/d/share"), .object(["args": .array([.string("/Users/d/share/tools/x.js")])]),
            "collapsing undoes what expanding did")
        XCTAssertEqual(Placeholder.collapseDirectory(in: config, directory: ""), config)
    }
}
